File: internal/rules/no_loss_of_precision/no_loss_of_precision_test.go
package no_loss_of_precision

import (
	"github.com/web-infra-dev/rslint/internal/rule_tester"
	"github.com/web-infra-dev/rslint/internal/rules/fixtures"
	"testing"
)

func TestNoLossOfPrecisionRule(t *testing.T) {
	rule_tester.RunRuleTester(
		fixtures.GetRootDir(),
		"tsconfig.json",
		t,
		&NoLossOfPrecisionRule,
		[]rule_tester.ValidTestCase{
			// Simple numbers
			{Code: `var x = 12345;`},
			{Code: `var x = 123.456;`},
			{Code: `var x = -123.456;`},
			{Code: `var x = 0;`},

			// Scientific notation
			{Code: `var x = 123e34;`},
			{Code: `var x = 123e-34;`},
			{Code: `var x = 12.3e-34;`},

			// MAX_SAFE_INTEGER
			{Code: `var x = 9007199254740991;`},

			// Large safe integers
			{Code: `var x = 12300000000000000000000000;`},

			// Small decimals
			{Code: `var x = 0.00000000000000000000000123;`},

			// Binary, octal, hex
			{Code: `var x = 0b11111111111111111111111111111111111111111111111111111;`},
			{Code: `var x = 0o377777777777777777;`},
			{Code: `var x = 0x1FFFFFFFFFFFFF;`},

			// Non-numeric
			{Code: `var x = true;`},
			{Code: `var x = 'abc';`},
			{Code: `var x = null;`},
		},
		[]rule_tester.InvalidTestCase{
			// Integers with precision loss
			{
				Code: `var x = 9007199254740993;`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "noLossOfPrecision"},
				},
			},
			{
				Code: `var x = 5123000000000000000000000000001;`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "noLossOfPrecision"},
				},
			},

			// Scientific notation with precision loss
			{
				Code: `var x = 9007199254740.993e3;`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "noLossOfPrecision"},
				},
			},
			{
				Code: `var x = 9.007199254740993e15;`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "noLossOfPrecision"},
				},
			},

			// Decimals with precision loss
			{
				Code: `var x = 900719.9254740994;`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "noLossOfPrecision"},
				},
			},

			// Hex with precision loss
			{
				Code: `var x = 0x20000000000001;`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "noLossOfPrecision"},
				},
			},

			// Binary with precision loss
			{
				Code: `var x = 0b100000000000000000000000000000000000000000000000000001;`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "noLossOfPrecision"},
				},
			},
		},
	)
}

File: packages/rslint-test-tools/tests/typescript-eslint/rules/no-loss-of-precision.test.ts
import { RuleTester } from '@typescript-eslint/rule-tester';



const ruleTester = new RuleTester();

ruleTester.run('no-loss-of-precision', {
  valid: [
    'const x = 12345;',
    'const x = 123.456;',
    'const x = -123.456;',
    'const x = 123_456;',
    'const x = 123_00_000_000_000_000_000_000_000;',
    'const x = 123.000_000_000_000_000_000_000_0;',
    'const x = 0;',
    'const x = 0.0;',
    'const x = 0.000000000000001;',
    'const x = -0.000000000000001;',
    'const x = 1234567890;',
    'const x = 9007199254740991;', // MAX_SAFE_INTEGER
    'const x = -9007199254740991;', // -MAX_SAFE_INTEGER
    'const x = 0x1fffffffffffff;', // MAX_SAFE_INTEGER in hex
    'const x = 0b11111111111111111111111111111111111111111111111111111;', // MAX_SAFE_INTEGER in binary
    'const x = 0o377777777777777777;', // MAX_SAFE_INTEGER in octal
    'const x = 123e34;',
    'const x = 123.456e78;',
  ],
  invalid: [
    {
      code: 'const x = 9007199254740993;',
      errors: [{ messageId: 'noLossOfPrecision' }],
    },
    {
      code: 'const x = 9_007_199_254_740_993;',
      errors: [{ messageId: 'noLossOfPrecision' }],
    },
    {
      code: 'const x = 9_007_199_254_740.993e3;',
      errors: [{ messageId: 'noLossOfPrecision' }],
    },
    {
      code: 'const x = 0b100_000_000_000_000_000_000_000_000_000_000_000_000_000_000_000_000_001;',
      errors: [{ messageId: 'noLossOfPrecision' }],
    },
    {
      code: 'const x = 5123000000000000000000000000001;',
      errors: [{ messageId: 'noLossOfPrecision' }],
    },
    {
      code: 'const x = -9007199254740993;',
      errors: [{ messageId: 'noLossOfPrecision' }],
    },
    {
      code: 'const x = 0x20000000000001;',
      errors: [{ messageId: 'noLossOfPrecision' }],
    },
    {
      code: 'const x = 0X20000000000001;',
      errors: [{ messageId: 'noLossOfPrecision' }],
    },
    {
      code: 'const x = 0o400000000000000001;',
      errors: [{ messageId: 'noLossOfPrecision' }],
    },
    {
      code: 'const x = 0O400000000000000001;',
      errors: [{ messageId: 'noLossOfPrecision' }],
    },
    {
      code: 'const x = 0b100000000000000000000000000000000000000000000000000001;',
      errors: [{ messageId: 'noLossOfPrecision' }],
    },
    {
      code: 'const x = 0B100000000000000000000000000000000000000000000000000001;',
      errors: [{ messageId: 'noLossOfPrecision' }],
    },
    {
      code: 'const x = 1.230000000000000000000000000000000000000000000000000000001;',
      errors: [{ messageId: 'noLossOfPrecision' }],
    },
    {
      code: 'const x = 1e999;',
      errors: [{ messageId: 'noLossOfPrecision' }],
    },
  ],
});
